core/src/impl/Kokkos_GraphImpl.hpp

```diff
@@ -0,0 +1,82 @@
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_KOKKOS_GRAPHIMPL_HPP
+#define KOKKOS_IMPL_KOKKOS_GRAPHIMPL_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <Kokkos_Core_fwd.hpp>
+
+#include <Kokkos_Concepts.hpp>  // is_execution_policy
+#include <impl/Kokkos_GraphImpl_fwd.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+template <class Policy>
+struct _add_graph_kernel_tag;
+
+template <template <class...> class PolicyTemplate, class... PolicyTraits>
+struct _add_graph_kernel_tag<PolicyTemplate<PolicyTraits...>> {
+  using type = PolicyTemplate<PolicyTraits..., IsGraphKernelTag>;
+};
+
+}  // end namespace Impl
+
+namespace Experimental {  // but not for users, so...
+
+template <class Policy>
+// requires ExecutionPolicy<Policy>
+auto require(Policy const& policy, Kokkos::Impl::KernelInGraphProperty) {
+  static_assert(Kokkos::is_execution_policy<Policy>::value,
+                "Internal implementation error!");
+  return typename Kokkos::Impl::_add_graph_kernel_tag<Policy>::type{policy};
+}
+
+}  // end namespace Experimental
```

Comment:
I would have expected this code in the `<Kokkos_ExecPolicy.hpp>` header because it already has a  `Experimental::require(Policy, ...)` overload.
Please elaborate why this new file might be a more appropriate place for it.

Reply:
Also since this function is not intended for user to call, should we place it in `Experimental::Impl::` or just `Impl::`?

Reply:
One of the goals of properties (and a number of similar mechanisms) is that you can hook the customization in the context that applies to the property, not in the context that defines the require customization point itself.  In other words, `require` is a (non-intrusive) operation on `KernelInGraphProperty` here, relating to graphs, and just like an overload of `operator+()` or whatever would go in the same file as the type rather than all in one place for a given project, the overload for `require` that pertains to a property should go with the property.

And no, it shouldn't go in `Impl` because we haven't made `require` into a proper customization point object yet so we can't check for ADL candidates in a context that doesn't include `Experimental::require` and all of that jazz. Right now, since we're just using function overloads, we have to put it in the exact namespace `Experimental` so that the qualified function call finds the right overload. This can be changed once we have proper properties infrastructure, but for now this should be fine.

Reply:
(I accept the answer to my first question)

> And no, it shouldn't go in Impl because we haven't made require into a proper customization point object yet so we can't check for ADL candidates in a context that doesn't include Experimental::require and all of that jazz. Right now, since we're just using function overloads, we have to put it in the exact namespace Experimental so that the qualified function call finds the right overload. This can be changed once we have proper properties infrastructure, but for now this should be fine.

I don't understand why you bring up ADL, this function is to be called from the Graph implementation and can be qualified it as appropriate there.  I suggested not to have it in `Experimental::` because the user is never expected to call it.

Reply:
I bring up ADL because the way the customization point objects (like `require` is supposed to be) are supposed to involve ADL lookup that doesn't find the customization point object itself. This way, all of the call sites can be `Experimental::require` even if the customization point implementations are not in that namespace (this is how much of ranges works, for instance). To avoid having to change the call sites once we fix this, I'm just putting the overload in the `Experimental` namespace for now.

Reply:
I think what Damien means is that we should consider NOT going through the public customization point, and even just use some other function impl_add_graphkerneltag_to_policy or whatever. For example now this totally private overload of require would show up in fancy IDEs when people look at what properties they can add to policies via require. If it were a different function (or I guess a different namespace) this overload would not show up.

Reply:
Actually I mean you could just call Kokkos::Impl::_add_graph_kernel_tag<Policy>::type{policy}; in the place where you right now call this public require function. Makes sense?